src/epics.c

```diff
@@ -0,0 +1,531 @@
+          "access",
+          ca_message(ret));
+    return (void *)1;
+  }
+
+  ca_replace_printf_handler(&printf_handler);
+
+  int i = 0;
+  for (i = 0; i < epics_plugin.pvs_num; ++i) {
+    struct pv *p = &epics_plugin.pvs[i];
+
+    ret = ca_create_channel(p->name, &connection_handler, p, 0, &p->id);
+    if (ret != ECA_NORMAL) {
+      ERROR("epics plugin: CA error %s occurred while trying to create channel "
+            "\"%s\"",
+            ca_message(ret), p->name);
+      ret = 1;
+      goto error;
+    }
+  }
+
+  while (epics_plugin.thread_loop != 0) {
+    ca_pend_event(2.0);
+  }
+
+error:
+  for (; i > 0; --i) {
+    struct pv *p = &epics_plugin.pvs[i - 1];
+
+    if (p->eid) {
+      ca_clear_subscription(p->eid);
+    }
+    ca_clear_channel(p->id);
+  }
+
+  ca_context_destroy();
+
+  return (void *)ret;
+}
+
+static int start_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  int ret = 0;
+  if (epics_plugin.thread_loop != 0) {
+    goto epics_unlock;
+  }
+
+  epics_plugin.thread_loop = 1;
+  ret = plugin_thread_create(&epics_plugin.thread_id, epics_thread, (void *)0,
+                             "epics");
+  if (ret != 0) {
+    epics_plugin.thread_loop = 0;
+    ERROR("epics plugin: Starting thread failed: %d", ret);
+
+    goto epics_unlock;
+  }
+
+  // FIXME: wait untill ca_context_create success
+
+epics_unlock:
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return ret;
+}
+
+static int stop_thread(void) {
+  pthread_mutex_lock(&epics_plugin.lock);
+
+  epics_plugin.thread_loop = 0;
+
+  pthread_mutex_unlock(&epics_plugin.lock);
+
+  return pthread_join(epics_plugin.thread_id, NULL);
+}
+
+static int epics_config_variable(oconfig_item_t *ci, struct pv *p) {
+
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong variable configuration");
+
+    return -1;
+  }
+
+  for (oconfig_item_t *c = ci->children; c != ci->children + ci->children_num;
+       ++c) {
+    if (strcasecmp(c->key, "Type") == 0) {
+      if (cf_util_get_string_buffer(c, p->value.type, sizeof(p->value.type)) !=
+          0) {
+
+        return -1;
+      }
+
+      sstrncpy(p->value.type_instance, p->name, sizeof(p->value.type_instance));
+    } else {
+      ERROR(
+          "epics plugin: Unknown configuration key \"%s\" for variable \"%s\"",
+          c->key, p->name);
+
+      return -1;
+    }
+  }
+
+  sstrncpy(p->value.plugin, "epics", sizeof(p->value.plugin));
+
+  return 0;
+}
+
+static int epics_config_label(oconfig_item_t *ci, struct pv *p) {
+  if (cf_util_get_string(ci, &p->name) != 0 || p->name == NULL) {
+    ERROR("epics plugin: Wrong label configuration");
+
+    return -1;
+  }
+
+  p->is_label = 1;
+
+  return 0;
+}
+
+static int epics_config(oconfig_item_t *ci) {
+  if (ci->children_num == 0) {
+    ERROR("epics plugin: No variables are specified");
+
+    return -1;
+  }
+
+  epics_plugin.pvs = calloc(ci->children_num, sizeof(struct pv));
```

Comment:
This should use `realloc` to allocate the memory, because the config callback can be called multiple times.

```c
struct pv *pvs = realloc(epics_plugin.pvs,
                         sizeof(*epics_plugin.pvs) * (epics_plugin.pvs_num + ci->children_num));
if (pvs == NULL) {
  return ENOMEM;
}
epics_plugin.pvs = pvs;
pvs = epics_plugin.pvs + epics_plugin.pvs_num;
memset(pvs, 0, sizeof(*pvs) * ci->children_num);
/* initialize elements in pvs[0 .. ci->children_num-1] here */
epics_plugin.pvs_num += ci->children_num;
```

Reply:
I didn't know about it. When does it happen? And what should be expected behavior when the childrens are different?

Reply:
Yeah, don't worry, this is not documented well and if it were, I think it's unreasonable to expect people to read the docs in this detail.

The callback is called multiple times when there are multiple `<Plugin epics>` blocks.